include/unifex/task.hpp

```diff
@@ -175,41 +175,52 @@ struct _promise {
     }
 
     template <typename Value>
-    decltype(auto) await_transform(Value&& value)
-        // TODO fix this noexcept
-        noexcept(is_nothrow_callable_v<tag_t<unifex::await_transform>, type&, Value>) {
-      if constexpr (derived_from<remove_cvref_t<Value>, _task_base> ||
-          blocking_kind::always_inline == cblocking<Value>()) {
+    decltype(auto) await_transform(Value&& value) {
+      if constexpr (derived_from<remove_cvref_t<Value>, _task_base>) {
         // We are co_await-ing a unifex::task or something that completes inline. We don't
         // need an additional transition.
         return unifex::await_transform(*this, (Value&&) value);
       } else if constexpr (tag_invocable<tag_t<unifex::await_transform>, type&, Value>
           || detail::_awaitable<Value>) {
         // Either await_transform has been customized or Value is an awaitable. Either
         // way, we can dispatch to the await_transform CPO, then insert a transition back
-        // to the correct execution context.
-        return unifex::await_transform(
-            *this,
-            typed_via(
-                as_sender(unifex::await_transform(*this, (Value&&) value)),
-                this->sched_));
+        // to the correct execution context if necessary.
+        return transform_awaitable_(unifex::await_transform(*this, (Value&&) value));
       } else if constexpr (unifex::sender<Value>) {
-        // If we are co_await'ing a sender that is the result of calling schedule,
-        // do something special
-        if constexpr (is_sender_for_v<remove_cvref_t<Value>, schedule>) {
-          return transform_schedule_sender_((Value&&) value);
-        } else {
-          // Otherwise, append a transition to the correct execution context and wrap the
-          // result in an awaiter:
-          return unifex::await_transform(*this, typed_via((Value&&) value, this->sched_));
-        }
+        return transform_sender_((Value&&) value);
       } else {
         // Otherwise, we don't know how to await this type. Just return it and let the
         // compiler issue a diagnostic.
         return (Value&&) value;
       }
     }
 
+    template <typename Awaitable>
+    decltype(auto) transform_awaitable_(Awaitable&& awaitable) {
+      if constexpr (blocking_kind::always_inline == cblocking<Awaitable>()) {
+        return (Awaitable&&) awaitable;
+      } else {
+        return unifex::await_transform(
+            *this,
+            typed_via(as_sender((Awaitable&&) awaitable), this->sched_));
```

Comment:
This obviously works, and, given the time pressure, we should probably go ahead with it, but is there any unnecessary overhead in converting an awaitable to a sender and back again?

Reply:
Agreed. I would like to write a `via` for awaitables, but this works for now.